src/main/scala/is/hail/expr/Relational.scala

```diff
@@ -0,0 +1,378 @@
+
+  def genotypeType: Type = metadata.genotypeSignature
+
+  def typ = TStruct(
+    "v" -> TVariant,
+    "va" -> vaType,
+    "s" -> TString,
+    "sa" -> saType,
+    "g" -> genotypeType)
+
+  def sampleEC: EvalContext = {
+    val aggregationST = Map(
+      "global" -> (0, globalType),
+      "s" -> (1, TString),
+      "sa" -> (2, saType),
+      "g" -> (3, genotypeType),
+      "v" -> (4, TVariant),
+      "va" -> (5, vaType))
+    EvalContext(Map(
+      "global" -> (0, globalType),
+      "s" -> (1, TString),
+      "sa" -> (2, saType),
+      "gs" -> (3, TAggregable(genotypeType, aggregationST))))
+  }
+
+  def variantEC: EvalContext = {
+    val aggregationST = Map(
+      "global" -> (0, globalType),
+      "v" -> (1, TVariant),
+      "va" -> (2, vaType),
+      "g" -> (3, genotypeType),
+      "s" -> (4, TString),
+      "sa" -> (5, saType))
+    EvalContext(Map(
+      "global" -> (0, globalType),
+      "v" -> (1, TVariant),
+      "va" -> (2, vaType),
+      "gs" -> (3, TAggregable(genotypeType, aggregationST))))
+  }
+}
+
+object NewAST {
+  def genericRewriteTopDown(ast: NewAST, rule: PartialFunction[NewAST, NewAST]): NewAST = {
+    def rewrite(ast: NewAST): NewAST = {
+      rule.lift(ast) match {
+        case Some(newAST) if newAST ne ast =>
+          rewrite(newAST)
+        case None =>
+          val newChildren = ast.children.map(rewrite)
+          if ((ast.children, newChildren).zipped.forall(_ eq _))
+            ast
+          else
+            ast.copy(newChildren)
+      }
+    }
+
+    rewrite(ast)
+  }
+
+  def rewriteTopDown[T](ast: MatrixAST[T], rule: PartialFunction[NewAST, NewAST]): MatrixAST[T] =
+    genericRewriteTopDown(ast, rule).asInstanceOf[MatrixAST[T]]
+
+  def rewriteTopDown[T](ast: KeyTableAST, rule: PartialFunction[NewAST, NewAST]): KeyTableAST =
+    genericRewriteTopDown(ast, rule).asInstanceOf[KeyTableAST]
+
+  def genericRewriteBottomUp(ast: NewAST, rule: PartialFunction[NewAST, NewAST]): NewAST = {
+    def rewrite(ast: NewAST): NewAST = {
+      val newChildren = ast.children.map(rewrite)
+
+      // only recons if necessary
+      val rewritten =
+        if ((ast.children, newChildren).zipped.forall(_ eq _))
+          ast
+        else
+          ast.copy(newChildren)
+
+      rule.lift(rewritten) match {
+        case Some(newAST) if newAST ne rewritten =>
+          rewrite(newAST)
+        case None =>
+          rewritten
+      }
+    }
+
+    rewrite(ast)
+  }
+  
+  def rewriteBottomUp[T](ast: MatrixAST[T], rule: PartialFunction[NewAST, NewAST]): MatrixAST[T] =
+    genericRewriteBottomUp(ast, rule).asInstanceOf[MatrixAST[T]]
+
+  def rewriteBottomUp[T](ast: KeyTableAST, rule: PartialFunction[NewAST, NewAST]): KeyTableAST =
+    genericRewriteBottomUp(ast, rule).asInstanceOf[KeyTableAST]
+}
+
+abstract class NewAST {
+  def children: IndexedSeq[NewAST]
+
+  def copy(newChildren: IndexedSeq[NewAST]): NewAST
+
+  def mapChildren(f: (NewAST) => NewAST): NewAST = {
```

Comment:
When you explicit type recursion from `Tree` you get two different functors, one on each type parameter.

```scala
sealed trait Tree[T]
final case class Leaf[T](value: T) extends Tree[T] {}
final case class Branch[T](left: Tree[T], right: Tree[T]) extends Tree [T] {}

object TreeIsFunctor {
  def map[T, U](t: Tree[T])(f: T => U): Tree[U] = t match {
    case Leaf(v) => Leaf(f(v))
    case Branch(l, r) => Branch(map(l)(f), map(r)(f))
  }
}
```

```scala
sealed trait Tree[T, R]
final case class Leaf[T, R](value: T) extends Tree[T, R] {}
final case class Branch[T, R](left: R, right: R) extends Tree[T, R] {}

object TreeIsFunctorInR {
  def map[T, R, S](t: Tree[T, R])(f: R => U): Tree[T, U] = t match {
    case Leaf(v) => Leaf(v)
    case Branch(l, r) => Branch(f(l), f(r))
  }
}
```

The latter is less obviously useful, especially because the type of tree of arbitrary depth cannot be written without a fix-point operator.

I'm fine with `mapChildren` too :)